app/src/org/commcare/models/database/app/AppDatabaseUpgrader.java

```diff
@@ -200,6 +215,99 @@ private boolean upgradeSevenEight(SQLiteDatabase db) {
         }
     }
 
+    // Migrate records form FormProvider and InstanceProvider to new FormDefRecord and FormRecord respectively
+    private boolean upgradeEightNine(SQLiteDatabase db) {
+        db.beginTransaction();
+        try {
+            upgradeXFormAndroidInstallerV1(db);
+
+            // Create FormDef table
+            TableBuilder builder = new TableBuilder(FormDefRecord.class);
+            db.execSQL(builder.getTableCreateString());
+
+            migrateFormProvier(db);
+            db.setTransactionSuccessful();
+            return true;
+        } finally {
+            db.endTransaction();
+        }
+    }
+
+    // migrate formProvider entries to db
+    private void migrateFormProvier(SQLiteDatabase db) {
+        Cursor cursor = null;
+        try {
+            cursor = context.getContentResolver().query(FormsProviderAPI.FormsColumns.CONTENT_URI, null, null, null, null);
+            if (cursor != null && cursor.getCount() > 0) {
+                SqlStorage<FormDefRecord> formDefRecordStorage = new SqlStorage<>(
+                        FormDefRecord.STORAGE_KEY,
+                        FormDefRecord.class,
+                        new ConcreteAndroidDbHelper(context, db));
+                while (cursor.moveToNext()) {
+                    FormDefRecord formDefRecord = new FormDefRecord(cursor);
+                    formDefRecord.save(formDefRecordStorage);
+                }
+            }
+        } finally {
+            safeCloseCursor(cursor);
+        }
+
+        // Delete migrated entries
+        context.getContentResolver().delete(FormsProviderAPI.FormsColumns.CONTENT_URI, null, null);
```

Comment:
So one quick point:

Errors in this operation can't be rolled back, and also aren't material to the future functionality of the app, since this CR will be ignored moving forward.

I think for those reasons that moving this op _outside_ of the v8->v9 Migrater makes a bit more sense. 

Right now, it's:

```
transaction[
1) add table
2) move data
3) delete old data
]
```
and if "delete old data" deletes _half_ of the data before failing, the first two ops will be undone, but the delete old data op _won't be_, which will put the phone in an irrecoverable state. If instead #3 was treated as a "best effort" op

```
transaction[
1) add table
2) move data
]
try{
3) delete old data
} catch(){
//oops... do nothing.
}
```
"Half of a failure" in step three doesn't result in an irrecoverable app.

Reply:
Ahh never realised deletes are not atomic. I think one other way to do this is wrap the delete call [here](https://github.com/dimagi/commcare-android/blob/mergeFormInsanceProviders/app/src/org/commcare/provider/InstanceProvider.java#L163) in a transaction. What do you think ?

Reply:
I *think* that would make everything here work reliably, as long as the transaction for this code is halted _immediately_ after the delete succeeds. I'm still a bit hesitant to introduce semantics where the ContentProvider.delete() call can succeed (in whole or in part) while the migrate transaction is still open, just as a matter of most-importance.

Like, if the line of code that says "commit the ContentProvider DB delete transaction" executes and then the phone battery dies, the next time the app spins up, our DB will roll back the migrate transaction and the content provider data will be gone. 

If we are super concerned with the ContentProvider delete occurring, I'd rather we separate those transactional semantics from the semantics of getting the data **into** this db. It's ok for the app to crash if the delete is failing (if that's a real bug that we should be catching and fixing), but I want to separate that condition from this transaction, since there's no way to use SQL transactions in a way which will Atomically link loading the new DB with the ContentProvider data, and deleting the data in the content provider DB. We'd need that to come from a separate source.

Reply:
Hmm I see the concerns. Will make the changes accordingly.